rfd/xxxx-automatic-azure-server-discovery.md

```diff
@@ -0,0 +1,217 @@
+      "teleport.dev/subscriptionId": "88888888-8888-8888-8888-888888888888"
+    }
+  },
+  "spec": {
+    "public_addr": "...",
+    "hostname": "azurexyz"
+  }
+}
+```
+
+### Agent installation
+
+In order to install the Teleport agent on Azure virtual machines, Teleport will serve an
+install script at `/webapi/scripts/{installer-resource-name}`. Installer scripts will
+be editable as a resource.
+
+Example resource script:
+```yaml
+kind: installer
+metadata:
+  name: "installer" # default value
+spec:
+  # shell script that will be downloaded and run by the virtual machine
+  script: |
+    #!/bin/sh
+    curl https://.../teleport-pubkey.asc ...
+    echo "deb [signed-by=... stable main" | tee ... > /dev/null
+    apt-get update
+    apt-get install teleport
+    teleport node configure --auth-agent=... --join-method=azure --token-name=azure-token
+  # Any resource in Teleport can automatically expire.
+  expires: 0001-01-01T00:00:00Z
+```
+
+Unless overridden by a user, a default teleport installer command will be
+generated that is appropriate for the current running version and operating
+system initially supporting DEB and RPM based distros that Teleport already
+provides packages for.
+
+To run commands, the agent's service principal will require the
+`Microsoft.Compute/virtualMachines/runCommand/action` permission.
+
+#### Join method
+
+Azure auto discovery will use the [Azure join method](https://github.com/gravitational/teleport/pull/19172)
+once it's implemented.
+
+#### Action vs Managed run commands
+
+Azure virtual machines can run scripts via either [Action Run Commands](https://learn.microsoft.com/en-us/azure/virtual-machines/linux/run-command)
+or [Managed Run Commands](https://learn.microsoft.com/en-us/azure/virtual-machines/linux/run-command-managed).
+Managed Run Commands have a few advantages over Action Run Commands, including:
+
+- Cancelable scripts
+- Finer control over timeouts
+- Progress reporting
+- Custom command IDs (this could make run commands more like SSM in AWS)
+
+Azure recommends Managed Run Commands for non-trivial installation,
+but they are still in Preview, so Teleport will use Action Run Commands. We
+may consider switching to Managed Run Commands when they are fully released.
+
+### teleport.yaml generation
+
+The `teleport node configure` subcommand will be used to generate a
+new /etc/teleport.yaml file:
+```sh
+teleport node configure
+    --auth-server=auth-server.example.com [auth server that is being connected to]
+    --token="$1" # passed via parameter from run-command
+    --labels="teleport.dev/subscriptionId=${SUBSCRIPTION},\
+    teleport.dev/resource-group=${RESOURCE_GROUP},\
+    teleport.dev/region=${REGION},\
+    teleport.dev/discovered-node=yes,\
+    teleport.dev/discovered-by=$2,\
+    teleport.dev/origin=cloud" # sourced from instance metadata
+```
+
+This will generate a file with the following contents:
+```yaml
+teleport:
+  nodename: "${SUBSCRIPTION}-${VM_ID}"
```

Comment:
Is this actually parameterized/templated somehow or are you using ${} syntax just to indicate that this will use sub/VM IDs?

Reply:
Just to indicate those values. I've changed the syntax to make that clearer.